### PR TITLE
Issue #12720: Make action to trigger all release steps

### LINKED ISSUE #12720: Make action to trigger all release steps
so we have all steps/phases of release automated but still required to be manually executed sequentially.
https://github.com/checkstyle/contribution/blob/master/releasenotes-builder/diagrams-manual-trigger.md

Ideally all of them should be auto triggered by themself.
One of the approach is: each action create git tag, other action triggered by tag creation.
https://github.com/checkstyle/contribution/blob/master/releasenotes-builder/diagrams.md

May be there is already some other way to execution actions one after another.

### PR DESCRIPTION
**Pull request by stoyanK7**
Resolves #12720

Continues from https://github.com/checkstyle/checkstyle/pull/12733. I can not reopen PR since it was from my `master` branch and I force-pushed. Probably there is a way to fix that but it is faster to open this one.

The first two commits are just for my convenience. We will remove them before merging. Look at commit `Issue #12720: Make action to trigger all release steps` to see the relevant changes to the issue.

Success run: https://github.com/stoyanK7/checkstyle/actions/runs/4177134639
Tags at my repo: https://github.com/stoyanK7/checkstyle/tags
Milestones: https://github.com/stoyanK7/checkstyle/milestones?state=closed
Commits: https://github.com/stoyanK7/checkstyle/commits/master
Sourceforge: https://stoyank7-checkstyle.sourceforge.io/

### CODE REVIEW

**Comment by rnveach** [comment]
> I can not reopen PR since it was from my master branch and I force-pushed. Probably there is a way to fix that but it is faster to open this one

Never use your own master branch. That is the easiest way to fix it. :)

File: .github/workflows/copy-site-to-sourceforge.yml [hunk 1]
```diff
@@ -6,17 +6,23 @@ name: "R: Copy Site to SourceForge"
...
+    inputs:
+      version:
+        type: string
+        required: true
+
 
-concurrency:
-  group: ${{ github.workflow }}-${{ github.event.pull_request.number || github.ref }}
-  cancel-in-progress: true
+#concurrency:
```
Comment (stoyanK7) [inline .github/workflows/copy-site-to-sourceforge.yml:19]: There is this small issue left with `concurrency`. Without removing it, we get:
https://github.com/stoyanK7/checkstyle/actions/runs/4177134639
Screenshot from 2023-02-14 20-39-15

All jobs who have to run in parallel are cancelled except one which is left to run.

Comment (stoyanK7) [reply .github/workflows/copy-site-to-sourceforge.yml:19]: I am currently trying to understand how this concurrency works and what we could do here.

Comment (nrmancuso) [reply .github/workflows/copy-site-to-sourceforge.yml:19]: Looks like you need to adjust concurrency settings, set cancel in progress to false

Comment (nrmancuso) [reply .github/workflows/copy-site-to-sourceforge.yml:19]: @stoyanK7 just use some grouping that would make each job have its own group name, check docs on this for job number or something

Comment (nrmancuso) [reply .github/workflows/copy-site-to-sourceforge.yml:19]: @stoyanK7 looks like you could do ${{ job.container.id }} to get a unique group for each job

File: .github/workflows/bump-version-and-update-milestone.yml [hunk 2]
```diff
@@ -23,7 +28,7 @@ jobs:
     runs-on: ubuntu-latest
     steps:
       - run: |
-          echo "Latest Version: ${{ github.event.inputs.version }}"
+          echo "Latest Version: ${{ inputs.version }}"
```
Comment (romani) [inline .github/workflows/bump-version-and-update-milestone.yml:31]: Is it possible to move such update on input usage in Separate PR, to let me merge it sooner and test in manual execution of each action in next release.
It is huge milestone for us that release is fully in CI even there are few actions to run.

PR with workflow will take a bit more time to accept.

**Review by romani — CHANGES_REQUESTED** [review]
Items

File: .github/workflows/bump-version-and-update-milestone.yml (hunk 2 above)
Comment (romani) [reply .github/workflows/bump-version-and-update-milestone.yml:31]: You can reference same issue in commit, so it is pre-step fore functional PR.

Comment (stoyanK7) [reply .github/workflows/bump-version-and-update-milestone.yml:31]: On it.

Comment (stoyanK7) [reply .github/workflows/bump-version-and-update-milestone.yml:31]: Done.

File: .github/workflows/maven-release-perform.yml [hunk 3]
```diff
@@ -31,15 +41,24 @@ jobs:
...
+          gpg-private-key: ${{ secrets.GPG_KEY }}
+          gpg-passphrase: GPG_PASSPHRASE
+      - name: Download xmlstarlet
+        run: |
+          sudo apt-get install xmlstarlet
+      - name: Setup SSH
+        uses: webfactory/ssh-agent@v0.7.0
+        with:
+          ssh-private-key: ${{ secrets.SSH_PRIVATE_KEY }}
+      - name: Run Shell Script
```
Comment (stoyanK7) [inline .github/workflows/maven-release-perform.yml:58]: I will open another PR for this too.

Comment (stoyanK7) [reply .github/workflows/maven-release-perform.yml:58]: Done.

File: .ci/create-github-release.sh [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+  exit 1
+fi
+
+TARGET_VERSION=$1
+echo TARGET_VERSION="$TARGET_VERSION"
+
+checkForVariable "GITHUB_TOKEN"
+
+curl -i -H "Authorization: token $GITHUB_TOKEN" \
+  -d "{ \"tag_name\": \"checkstyle-$TARGET_VERSION\", \
```
Comment (nrmancuso) [inline .ci/create-github-release.sh:19]: Please use heredoc to make json body more readable/maintainable. See an example in the project at https://github.com/checkstyle/checkstyle/blob/a359c4121aa15aad1cd147e0ebfc1528bfd82bae/.ci/update-github-page.sh#L80, note that you can also just cat the heredoc in a subshell and assign it to some variable, like `X=$(cat EOF {"myjson":{}} EOF)`.

File: .github/workflows/release.yml [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+        required: true
+
+jobs:
+  update-sources:
+    uses: ./.github/workflows/update-sources.yml
+    with:
+      version: ${{ inputs.version }}
+    permissions:
+      contents: write
+  maven-prepare:
```
Comment (nrmancuso) [inline .github/workflows/release.yml:20]: Please separate jobs by an empty line

**Review by nrmancuso — CHANGES_REQUESTED** [review]
Items:

File: .github/workflows/copy-site-to-sourceforge.yml (hunk 1 above)
Comment (stoyanK7) [reply .github/workflows/copy-site-to-sourceforge.yml:19]: I assume we want to use `job.container.id` only when the event trigger is `workflow_call`(when it is reused/called from another workflow) and for the rest of the time leave as it is right now, correct?

Comment (stoyanK7) [reply .github/workflows/copy-site-to-sourceforge.yml:19]: @nrmancuso Using this variable on a job level concurrent does not work:
https://github.com/stoyanK7/checkstyle/actions/runs/4192845855
```
Invalid workflow file: .github/workflows/release.yml#L56
The workflow is not valid. .github/workflows/release.yml (Line: 56, Col: 14): Unrecognized named-value: 'job'. Located at position 1 within expression: job.container.id .github/workflows/release.yml (Line: 65, Col: 17): Unrecognized named-value: 'job'. Located at position 1 within expression: job.container.id

```
```yml
  new-milestone-and-issue-in-other-repo:
    uses: ./.github/workflows/new-milestone-and-issue-in-other-repo.yml
    needs: create-github-release
    with:
      version: ${{ inputs.version }}
    permissions:
      contents: write
      issues: write
    concurrency:
      group: ${{ job.container.id }}
```

Comment (stoyanK7) [reply .github/workflows/copy-site-to-sourceforge.yml:19]: This is not allowed as well

```yml
concurrency:
  - group: ${{ github.run_id }}
    if: ${{ github.event_name == 'workflow_call' }}
  - group: ${{ github.workflow }}-${{ github.event.pull_request.number || github.ref }}
    if: ${{ github.event_name == 'workflow_dispatch' }}
    cancel-in-progress: true
```

Comment (stoyanK7) [reply .github/workflows/copy-site-to-sourceforge.yml:19]:
```yml
  new-milestone-and-issue-in-other-repo:
    ...
    concurrency:
      group: a1
      cancel-in-progress: true

  publish-releasenotes-outside:
    ...
    concurrency:
      group: a2
      cancel-in-progress: true
```
still cancels them https://github.com/stoyanK7/checkstyle/actions/runs/4193090241

File: .github/workflows/create-github-release.yml [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+concurrency:
+  group: ${{ github.workflow }}-${{ github.event.pull_request.number || github.ref }}
+  cancel-in-progress: true
+
+permissions:
+  contents: write
+
+jobs:
+  publish:
+    name: Create GitHub release for ${{ inputs.version }}
```
Comment (nrmancuso) [inline .github/workflows/create-github-release.yml:27]: @stoyanK7 please point to where we are able to resolve name for jobs with context variables in the name; I do not see it here: https://github.com/stoyanK7/checkstyle/actions/runs/4193090241 . Most CI I have dealt with are not able to resolve variables in job/workflow names

File: .github/workflows/copy-site-to-sourceforge.yml (hunk 1 above)
Comment (nrmancuso) [reply .github/workflows/copy-site-to-sourceforge.yml:19]: > @nrmancuso Using this variable on a job level concurrent does not work:

Yeah, it makes sense that this variable is not available until we start contianer.

> still cancels them https://github.com/stoyanK7/checkstyle/actions/runs/4193090241

I am not sure about this; it looks like one was cancelled, but the other two jobs were skipped.

File: .github/workflows/create-github-release.yml (hunk 6 above)
Comment (stoyanK7) [reply .github/workflows/create-github-release.yml:27]: image

Is this what you mean? You can see the name on the top.

Comment (nrmancuso) [reply .github/workflows/create-github-release.yml:27]: Yes, I see that if job is successful we can resolve variable: https://github.com/stoyanK7/checkstyle/actions/runs/4177134639

Strange that we do not when it fails: https://github.com/stoyanK7/checkstyle/actions/runs/4193090241

Either way, if it works in successful run it is worth keeping.

Comment (nrmancuso) [reply .github/workflows/create-github-release.yml:27]: Example of failure to resolve:
image

File: .github/workflows/copy-site-to-sourceforge.yml (hunk 1 above)
Comment (nrmancuso) [reply .github/workflows/copy-site-to-sourceforge.yml:19]: @stoyanK7 I have a theory, but we would need to investigate; I notice that most/all workflows have a space in the name. I wonder if this is causing an issue (perhaps workflow name is getting dropped/truncated/whatever). Otherwise, there should be no issue since we have a group naming scheme like: `${{ github.workflow }}-${{ github.event.pull_request.number || github.ref }}`, so example would be `R: Copy Site to SourceForge-12741`

So, each concurrency group would be unique to that workflow. But - if we drop everything after the first space, then we get `R:-12741`, and all workflows would be trying to run in this group.

Comment (nrmancuso) [reply .github/workflows/copy-site-to-sourceforge.yml:19]: Please try https://github.com/checkstyle/checkstyle/pull/12741/commits/74d9ab6c73b9f815bb2b15df0b94f0f72367ccf7#r1108301551 again with `cancel-in-progress: false`, workflows should be queued in this case, but not cancelled.

File: .github/workflows/bump-version-and-update-milestone.yml (hunk 2 above)
Comment (romani) [reply .github/workflows/bump-version-and-update-milestone.yml:31]: merged, please update this Pr

File: .github/workflows/copy-site-to-sourceforge.yml (hunk 1 above)
Comment (nrmancuso) [reply .github/workflows/copy-site-to-sourceforge.yml:19]: @stoyanK7 OK,  here is the problem: https://docs.github.com/en/actions/using-workflows/reusing-workflows#:~:text=the%20github%20context%20is%20always%20associated%20with%20the%20caller%20workflow.

> the github context is always associated with the caller workflow.

We need to use some other context/means for naming

Comment (stoyanK7) [reply .github/workflows/copy-site-to-sourceforge.yml:19]: > Please try https://github.com/checkstyle/checkstyle/commit/74d9ab6c73b9f815bb2b15df0b94f0f72367ccf7#r1108301551 again with cancel-in-progress: false, workflows should be queued in this case, but not cancelled if this is a valid grouping.

Screenshot from 2023-02-17 16-10-10

https://github.com/stoyanK7/checkstyle/actions/runs/4205004190

Comment (stoyanK7) [reply .github/workflows/copy-site-to-sourceforge.yml:19]: > You could also pass the name of the workflow itself as a parameter to the child workflow.

You mean like as an input?
```yml
... release.yml
  new-milestone-and-issue-in-other-repo:
    uses: ./.github/workflows/new-milestone-and-issue-in-other-repo.yml
    with:
      version: ${{ inputs.version }}
      concurrency-group: new-milestone
```

and then
```yml
... new-milestone-and-issue-in-other-repo.yml
on:
  workflow_call:
    inputs:
      version:
        type: string
        required: true
    concurrency-group:
        type: string
        required: true
        default: ""
```

This is what I assume you mean.

Comment (stoyanK7) [reply .github/workflows/copy-site-to-sourceforge.yml:19]: Here is the result. It works. They ran simultaneously

Screenshot from 2023-02-17 16-24-41

https://github.com/stoyanK7/checkstyle/actions/runs/4205128732

File: .github/workflows/release.yml (hunk 5 above)
Comment (stoyanK7) [reply .github/workflows/release.yml:21]: Done.

File: .ci/create-github-release.sh (hunk 4 above)
Comment (stoyanK7) [reply .ci/create-github-release.sh]: Done.

File: .github/workflows/copy-site-to-sourceforge.yml (hunk 1 above)
Comment (stoyanK7) [reply .github/workflows/copy-site-to-sourceforge.yml]: Done. Pushed changes.

**Comment by stoyanK7** [comment]
@nrmancuso ping

**Comment by stoyanK7** [comment]
Removed unnecessary commits and left only what is to be changed/merged.

File: .ci/create-github-release.sh [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+  "body": "",
+  "draft": false,
+  "prerelease": false
+}
+EOF
+)
+
+curl -i -H "Authorization: token $PAT" \
+  -d "$BODY" \
+  -X POST https://api.github.com/repos/stoyanK7/checkstyle/releases
```
Comment (nrmancuso) [inline .ci/create-github-release.sh:32]: @stoyanK7 please fix url

**Review by nrmancuso — CHANGES_REQUESTED** [review]
Item:

File: .ci/create-github-release.sh (hunk 7 above)
Comment (stoyanK7) [reply .ci/create-github-release.sh]: That one slipped away. Done.

**Review by nrmancuso — APPROVED** [review]

File: .github/workflows/update-sources.yml [hunk 8]
```diff
@@ -23,6 +28,9 @@ jobs:
     steps:
       - name: Checkout the latest code
         uses: actions/checkout@v3
+        with:
+          fetch-depth: 0
+          ref: master
```
Comment (romani) [inline .github/workflows/update-sources.yml:33]: if this is just a speedup of clone execution, please move it to separate PR to merge sooner.

File: .github/workflows/update-github-io.yml [hunk 9]
```diff
@@ -9,9 +9,18 @@ on:
...
+        type: string
+        required: true
+      concurrency-group:
+        type: string
+        required: true
 
 concurrency:
-  group: ${{ github.workflow }}-${{ github.event.pull_request.number || github.ref }}
+  group: ${{ github.workflow }}
+    -${{ github.event.pull_request.number || github.ref }}-${{ inputs.concurrency-group }}
```
Comment (romani) [inline .github/workflows/update-github-io.yml:23]: sounds like not related optimization, move to separate PR to merge sooner.

File: .github/workflows/copy-site-to-sourceforge.yml [hunk 10]
```diff
@@ -6,17 +6,22 @@ name: "R: Copy Site to SourceForge"
 on:
   workflow_dispatch:
     inputs:
-      release-version:
+      version:
```
Comment (romani) [inline .github/workflows/copy-site-to-sourceforge.yml:9]: sounds like not related optimization, move to separate PR to merge sooner.

**Review by romani — CHANGES_REQUESTED** [review]
there are still bunch of improvements that can be merged unrelated to this issue, please move them out to merge sooner:

File: .ci/create-github-release.sh [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+checkForVariable "PAT"
+
+BODY=$(cat <<EOF
+{
+  "tag_name": "checkstyle-$TARGET_VERSION",
+  "target_commitish": "master",
+  "name": "",
+  "body": "",
+  "draft": false,
+  "prerelease": false
```
[the remark this answers is not in the document]
Comment (stoyanK7) [reply .ci/create-github-release.sh:25]: The purpose of this script is to create the release on GitHub. Nowhere in the release process are we doing that:
image


After it is created, `R: Close/Create Milestone, Upload '-all' jar, Create issues` and `R: Publish GitHub Page` search for that release by tag `checkstyle-<release target version>`
https://github.com/checkstyle/checkstyle/blob/fb9ff40ef7679342afcc6dc62cd059255c673dbb/.ci/update-github-page.sh#L36-L42
https://github.com/checkstyle/checkstyle/blob/fb9ff40ef7679342afcc6dc62cd059255c673dbb/.ci/upload-all-jar.sh#L23-L28

After both of these jobs update the release, this is how it looks:
image


Here is a successful run with the job that creates the GitHub release:
https://github.com/stoyanK7/checkstyle/actions/runs/4254777591

And here is a failing run when we do not use that job: https://github.com/stoyanK7/checkstyle/actions/runs/4254920912
 - https://github.com/stoyanK7/checkstyle/actions/runs/4254920912/jobs/7401938612#step:4:121
   ```
   TARGET_RELEASE_INDEX=
   PREVIOUS_RELEASE_INDEX=1
   START_REF=checkstyle-10.6.0
   END_REF=checkstyle-10.7.0
   ...
   Exception in thread "main" org.eclipse.jgit.errors.RevisionSyntaxException: Invalid ref name: checkstyle-10.7.0
   ```
 - https://github.com/stoyanK7/checkstyle/actions/runs/4254920912/jobs/7401938788#step:6:243
   ```
   HTTP/2 404 
   ...
   {"message":"Not Found","request_id":"1BC5:0714:6AADEC:889F39:63F79B70","documentation_url":"https://docs.github.com/rest"}Jar Published for 10.8.0
  ```

[the remark this answers is not in the document]
Comment (stoyanK7) [reply .ci/create-github-release.sh:25]: You were doing the same thing in `release.sh`:
https://github.com/checkstyle/checkstyle/blob/fb9ff40ef7679342afcc6dc62cd059255c673dbb/release.sh#L67-L74

File: .github/workflows/update-github-io.yml (hunk 9 above)
Comment (stoyanK7) [reply .github/workflows/update-github-io.yml:23]: It is related to the PR, we use `${{ inpuits.concurrency-group }}` which we get from the caller workflow(`release.yml`).

File: .github/workflows/copy-site-to-sourceforge.yml (hunk 10 above)
Comment (stoyanK7) [reply .github/workflows/copy-site-to-sourceforge.yml:9]: I'd say it is related too. We want to have the same release input name for both `workflow_dispatch` and `workflow_call` so we can target both with `${{ inputs.version }}`. Hence the change.

File: .github/workflows/update-sources.yml (hunk 8 above)
Comment (stoyanK7) [reply .github/workflows/update-sources.yml:33]: It is not a speedup of execution. This one in particular can be removed, but the rest of
```yml
        with:
          fetch-depth: 0
          ref: master
```
that I added are required when we have a job that is dependent on a previous job that commits and pushes to repo,

There is an issue for this at `actions/checkout`:https://github.com/actions/checkout/issues/439#issue-804387669

I use the solution proposed in the comments: https://github.com/actions/checkout/issues/439#issuecomment-965968956

File: .github/workflows/release.yml [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+name: "Release"
+on:
+  workflow_dispatch:
+    inputs:
+      version:
+        description: 'Target version without (-SNAPSHOT)'
+        required: true
+
+concurrency:
+  group: ${{ github.workflow }}-${{ github.event.pull_request.number || github.ref }}
```
Comment (stoyanK7) [inline .github/workflows/release.yml:14]: @nrmancuso We run into a concurrency deadlock here. Either we have to remove it or use a new one.

Concurrency problem: 
https://github.com/stoyanK7/checkstyle/actions/runs/4254614589
```
Release: .github#L1
Canceling since a deadlock for concurrency group 'Release-refs/heads/master' was detected between 'top level workflow' and 'update-sources'
```

What concurrency group do you suggest we use?

File: .github/workflows/update-github-io.yml (hunk 9 above)
Comment (stoyanK7) [reply .github/workflows/update-github-io.yml:23]: Moreover, this change was applied after discussion in this PR above: https://github.com/checkstyle/checkstyle/pull/12741#discussion_r1108528505

File: .github/workflows/release.yml (hunk 12 above)
Comment (nrmancuso) [reply .github/workflows/release.yml:14]: Remember that we pass `github` context to child workflows, this is probably causing the issue. Let's just remove all release workflows from the check at https://github.com/checkstyle/checkstyle/blob/d6d906980759598be045000fd32f3567d2b7c3d5/.ci/validation.sh#L1111

Comment (stoyanK7) [reply .github/workflows/release.yml:14]: > Remember that we pass github context to child workflows

I do remember that. My thought was - we can give `release.yml` some unique concurrency group so we can still take advantage of concurrency functionality. For example, `Caller-${{ github.workflow }}-${{ github.event.pull_request.number || github.ref }}`.

https://docs.github.com/en/actions/using-workflows/reusing-workflows#overview
> A workflow that uses another workflow is referred to as a "caller" workflow.

Comment (nrmancuso) [reply .github/workflows/release.yml:14]: Makes sense to me, let’s try it

File: .github/workflows/copy-site-to-sourceforge.yml (hunk 10 above)
Comment (stoyanK7) [reply .github/workflows/copy-site-to-sourceforge.yml:9]: Moved.

File: .ci/create-github-release.sh (hunk 11 above)
[the remark this answers is not in the document]
Comment (stoyanK7) [reply .ci/create-github-release.sh:25]: Moved to another PR.

[the remark this answers is not in the document]
Comment (romani) [reply .ci/create-github-release.sh:25]: @stoyanK7, so you mean it was simply lost/forgotten during split of single file to multiple ?

[the remark this answers is not in the document]
Comment (stoyanK7) [reply .ci/create-github-release.sh:25]: It seems like it. But strange that you did not have any issue with semi-automated release.

**Comment by stoyanK7** [comment]
Back on it now that some things are merged.

File: .github/workflows/copy-site-to-sourceforge.yml (hunk 10 above)
Comment (stoyanK7) [reply .github/workflows/copy-site-to-sourceforge.yml:9]: This one can be resolved.

File: .ci/create-github-release.sh (hunk 11 above)
[the remark this answers is not in the document]
Comment (stoyanK7) [reply .ci/create-github-release.sh:25]: That one can be resolved as well.

File: .github/workflows/release.yml (hunk 12 above)
Comment (stoyanK7) [reply .github/workflows/release.yml:14]: Works. All jobs are running.
https://github.com/stoyanK7/checkstyle/actions/runs/4274211452
Screenshot 2023-02-26 at 10-06-29 Release · stoyanK7_checkstyle@738cec2

**Review by romani — APPROVED** [review]
ok to merge

**Review by rnveach — APPROVED** [review]
I guess this ok, but I can't fully validate it.
